### ISSUE TITLE
Fully Programmatic Language Features

### ISSUE THREAD

**Issue by AVagueNumberOfHumans** [LuqueDaniel/vscode-language-renpy#11]
Would anyone be interested in expanding the VS Code Ren'py language support extension into a fully programmatic Ren'py debugger with me? I am tired of not having go-to-definition, but I am fairly new to Ren'py. 
References: 
* VS Code Programmatic Language Extension Features: 
  * https://code.visualstudio.com/api/language-extensions/programmatic-language-features
* Partially completed Renpy Text Editor: 
  * https://github.com/Dogeek/renpy-text-editor
* Ren'py support for Atom github repository: 
  * https://github.com/renpy/language-renpy
* Ren'py repository: 
  * https://github.com/renpy/renpy

**Comment by AVagueNumberOfHumans** [comment]
I also found a Ren'py debugger, https://github.com/Enerccio/librpydb 
currently being used in the partially completed Ren'py Text Editor: https://github.com/Dogeek/renpy-text-editor

**Comment by LuqueDaniel** [comment]
@AVagueNumberOfHumans Thank you for sharing those references! I would definitely like to implement these features. 👍

**Comment by AVagueNumberOfHumans** [comment]
Awesome! I've settled on the idea of building a Renpy language server for the extension, using this as a guide: https://medium.com/ballerina-techblog/implementing-a-language-server-how-hard-can-it-be-part-1-introduction-c915d2437076

**Comment by AVagueNumberOfHumans** [comment]
I just found a ton of examples: https://github.com/Microsoft/language-server-protocol/blob/gh-pages/_implementors/servers.md

**Comment by AVagueNumberOfHumans** [comment]
I'm starting a repo for the Ren'Py Language Server: https://github.com/AVagueNumberOfHumans/renpyls 
and writing it with the help of LSP4J: https://github.com/eclipse/lsp4j

**Comment by LuqueDaniel** [comment]
I don't have much experience with Java, but I'd like to help.

**Comment by rdurfee** [comment]
I updated the extension to implement several of the VS Code features. It uses the dump file Renpy creates when compiling (game/saves/navigation.json). It can detect when the dump file has changed (i.e., you ran the game or did a force compile), but it can't at this time trigger a compilation. (Renpy does have a command line option to re-build the dump file for a workspace folder however.)

I also tried using the sphinx project to extract the Renpy documented features, but ran into a problem here because most of the doc attributes are missing. I'm not sure how to continue with that feature.

I was able to implement:
* HoverProvider
* DefinitionProvider (go to Definition)
* DocumentSymbolProvider (the Outline view)
* CompletionItemProvider
* DocumentColorProvider

Hover and Definition aren't currently smart enough to know about context of what's hovered (i.e., if you hover of a symbol that's embedded in dialogue, it will show the hover).

I put this together over a few days and the code could use some clean up and a second look, but it could be a start, especially if we can get the documented RenPy docs.

Is this something you'd want to see in a PR to take a look at?
